File: el-lemby-platformer/windows/ElLemby.App/ResultScene.cs
using ElLemby.Core;

namespace ElLemby.App;

internal enum ResultKind
{
    StageClear,
    GameOver,
}

/// <summary>Shared end screen for «مبروك» and «انتهت اللعبة».</summary>
internal sealed class ResultScene : IScene
{
    private readonly ISceneHost _host;
    private readonly ResultKind _kind;
    private readonly int _timeBonus;

    public ResultScene(ISceneHost host, ResultKind kind, int timeBonus = 0)
    {
        _host = host;
        _kind = kind;
        _timeBonus = timeBonus;
    }

    public void Update(double dt, double now) { }

    public void KeyPressed(Keys key, double now)
    {
        switch (key)
        {
            case Keys.Space:
                _host.State.ResetRun();
                _host.Switch(new GameScene(_host));
                break;
            case Keys.Enter or Keys.Escape:
                _host.Switch(new TitleScene(_host));
                break;
            case Keys.M:
                Audio.ToggleMute();
                break;
        }
    }

    public void KeyReleased(Keys key, double now) { }

    public void Render(Graphics g, double now)
    {
        g.Clear(Palette.Night);
        GameState state = _host.State;

        var stats = new List<string>();
        if (_kind == ResultKind.StageClear)
        {
            Draw.TextCenter(g, L10n.StageClear, Draw.H1, Palette.GoldBrush, 240, 44);
            Draw.TextCenter(g, L10n.StageClearSub, Draw.Body, Palette.CreamBrush, 240, 78);
            stats.Add($"{L10n.TimeBonusLabel}: {L10n.Eastern(_timeBonus)}");

            // The happy couple, with a pulsing heart.
            g.DrawImage(Assets.Sprite("lemby_idle_0"), new Rectangle(240 - 60, 112, 40, 60));
            g.DrawImage(Assets.Sprite("nousa_0"), new Rectangle(240 + 20, 112, 40, 60));
            int beat = (int)(6 + 4 * Math.Abs(Math.Sin(now * 3.5)));
            g.DrawImage(Assets.Sprite("heart"),
                        new Rectangle(240 - beat, 100 - beat, beat * 2, beat * 2));
        }
        else
        {
            Draw.TextCenter(g, L10n.GameOver, Draw.H1, Palette.MaroonBrush, 240, 44);
            Draw.TextCenter(g, L10n.GameOverQuote, Draw.Body, Palette.CreamBrush, 240, 78);
            g.DrawImage(Assets.Sprite("lemby_hurt_0"), new Rectangle(240 - 20, 112, 40, 60));
        }

        stats.Add($"{L10n.MoneyLabel}: {L10n.Eastern(state.Money)}");
        stats.Add($"{L10n.ScoreLabel}: {L10n.Eastern(state.Score)}");
        stats.Add($"{L10n.HighScoreLabel}: {L10n.Eastern(state.HighScore)}");

        for (int i = 0; i < stats.Count; i++)
        {
            Draw.TextCenter(g, stats[i], Draw.Body, Palette.CreamBrush, 240, 184 + i * 17);
        }

        if ((int)(now / 0.6) % 2 == 0)
        {
            Draw.TextCenter(g, L10n.RetryHint, Draw.Tiny, Palette.CreamBrush, 240, 250);
        }
    }
}

File: el-lemby-platformer/Sources/ElLembyCore/Scenes/ResultScene.swift
import AppKit
import SpriteKit

enum ResultKind {
    case stageClear(timeBonus: Int)
    case gameOver
}

/// Shared end screen for both «مبروك» and «انتهت اللعبة».
final class ResultScene: SKScene {
    private let kind: ResultKind

    init(size: CGSize, kind: ResultKind) {
        self.kind = kind
        super.init(size: size)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) is not supported")
    }

    override var acceptsFirstResponder: Bool { true }

    override func didMove(to view: SKView) {
        backgroundColor = Palette.night
        let state = GameState.shared
        let centerX = size.width / 2

        let title = SKLabelNode(fontNamed: Fonts.arabicBold)
        title.fontSize = 30
        title.position = CGPoint(x: centerX, y: size.height - 78)
        addChild(title)

        let subtitle = SKLabelNode(fontNamed: Fonts.arabic)
        subtitle.fontSize = 13
        subtitle.fontColor = Palette.cream
        subtitle.position = CGPoint(x: centerX, y: size.height - 102)
        addChild(subtitle)

        var statLines: [String] = []
        switch kind {
        case let .stageClear(timeBonus):
            title.text = L10n.stageClear
            title.fontColor = Palette.gold
            subtitle.text = L10n.stageClearSub
            statLines.append("\(L10n.timeBonusLabel): \(L10n.eastern(timeBonus))")

            let couple = SKNode()
            let lemby = SKSpriteNode(texture: SpriteLoader.texture("lemby_idle_0"),
                                     size: CGSize(width: 16, height: 24))
            lemby.position = CGPoint(x: -16, y: 0)
            let nousa = SKSpriteNode(texture: SpriteLoader.texture("nousa_0"),
                                     size: CGSize(width: 16, height: 24))
            nousa.position = CGPoint(x: 16, y: 0)
            let heart = SKSpriteNode(texture: SpriteLoader.texture("heart"),
                                     size: CGSize(width: 8, height: 8))
            heart.position = CGPoint(x: 0, y: 18)
            couple.addChild(lemby)
            couple.addChild(nousa)
            couple.addChild(heart)
            couple.setScale(2.5)
            couple.position = CGPoint(x: centerX, y: size.height / 2 + 6)
            addChild(couple)
            heart.run(.repeatForever(.sequence([
                .scale(to: 1.35, duration: 0.4),
                .scale(to: 1.0, duration: 0.4),
            ])))
        case .gameOver:
            title.text = L10n.gameOver
            title.fontColor = Palette.maroon
            subtitle.text = L10n.gameOverQuote

            let lemby = SKSpriteNode(texture: SpriteLoader.texture("lemby_hurt_0"),
                                     size: CGSize(width: 16, height: 24))
            lemby.setScale(2.5)
            lemby.position = CGPoint(x: centerX, y: size.height / 2 + 6)
            addChild(lemby)
        }

        statLines.append("\(L10n.moneyLabel): \(L10n.eastern(state.money))")
        statLines.append("\(L10n.scoreLabel): \(L10n.eastern(state.score))")
        statLines.append("\(L10n.highScoreLabel): \(L10n.eastern(state.highScore))")

        for (i, line) in statLines.enumerated() {
            let label = SKLabelNode(fontNamed: Fonts.arabic)
            label.text = line
            label.fontSize = 12
            label.fontColor = Palette.cream
            label.position = CGPoint(x: centerX, y: 108 - CGFloat(i) * 18)
            addChild(label)
        }

        let hint = SKLabelNode(fontNamed: Fonts.arabic)
        hint.text = L10n.retryHint
        hint.fontSize = 10
        hint.fontColor = Palette.cream
        hint.alpha = 0.8
        hint.position = CGPoint(x: centerX, y: 22)
        addChild(hint)
        hint.run(.repeatForever(.sequence([
            .fadeAlpha(to: 0.3, duration: 0.6),
            .fadeAlpha(to: 0.8, duration: 0.6),
        ])))

        view.window?.makeFirstResponder(self)
    }

    override func keyDown(with event: NSEvent) {
        switch KeyCode(rawValue: event.keyCode) {
        case .space:
            SceneRouter.startGame(in: view)
        case .returnKey, .escape:
            SceneRouter.showTitle(in: view)
        case .m:
            SoundManager.shared.toggleMute()
        default:
            break
        }
    }
}
